[lldb-dap] Added "port" property to vscode "attach" command.

santhoshe447 opened the pull request:
Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuration to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote ".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
"version": "0.2.0",
"configurations": [
{
"name": "lldb-dap Debug",
"type": "lldb-dap",
"request": "attach",
"gdb-remote-port":1234,
"program": "${workspaceFolder}/a.out",
"args": [],
"stopOnEntry": false,
"cwd": "${workspaceFolder}",
"env": [],

    }
]
}

Commit:
[lldb][test] Add the ability to extract the variable value out of the summary.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuratiuon to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote <port>".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
	"version": "0.2.0",
    "configurations": [
        {
            "name": "lldb-dap Debug",
            "type": "lldb-dap",
            "request": "attach",
            "port":1234,
            "program": "${workspaceFolder}/a.out",
            "args": [],
            "stopOnEntry": false,
            "cwd": "${workspaceFolder}",
            "env": [],

        }
    ]
}

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.

Adding a "port" property to the VsCode "attach" command likely extends the functionality of the debugger configuration to allow attaching to a process using PID or PORT number.
Currently, the "Attach" configuration lets the user specify a pid. We tell the user to use the attachCommands property to run "gdb-remote ".
Followed the below conditions for "attach" command with "port" and "pid"
We should add a "port" property. If port is specified and pid is not, use that port to attach. If both port and pid are specified, return an error saying that the user can't specify both pid and port.

Ex - launch.json
{
"version": "0.2.0",
"configurations": [
{
"name": "lldb-dap Debug",
"type": "lldb-dap",
"request": "attach",
"port":1234,
"program": "${workspaceFolder}/a.out",
"args": [],
"stopOnEntry": false,
"cwd": "${workspaceFolder}",
"env": [],

    }
]
}

In this patch we have resolved code formatting issues and fixed "test_by_name" failure.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
Resolved code format issue.

Commit:
Resolved code format errors.

walter-erquinigo commented on lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py:
```diff
@@ -335,6 +335,37 @@ def cleanup():
                 response["success"], "attach failed (%s)" % (response["message"])
             )
 
+    def attach_by_port(
```
could you just extend the `def attach` function? Having a totally new entry point just adds more maintainance burden when doing refactors.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,137 @@
+
+class TestDAP_attachByPortNum(lldbdap_testcase.DAPTestCaseBase):
+    def runTargetProgramOnPort(self, port=None, program=None):
+        server_tool = "lldb-server"
```
lldb-server is available on linux, and gdbserver on mac, so you need to look for both depending on the situation.

@JDevlieghere , is there an SB API that we could use to get the path to the gdb-server that LLDB uses?

santhoshe447 replied:
I think "**lldbgdbserverutils.get_debugserver_exe()**" this function would help us to get gdbserver path.

walter-erquinigo replied:
that's a good idea!

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -762,9 +765,31 @@ void request_attach(const llvm::json::Object &request) {
+          (port != LLDB_INVALID_PORT_NUMBER)) {
+        // If both pid and port numbers are specified.
+        error.SetErrorString("The user can't specify both pid and port");
+      } else if ((port != LLDB_INVALID_PORT_NUMBER) && (port < UINT16_MAX)) {
```
Remove `(port < UINT16_MAX)` so that we get a meaningful error from the connection logic.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - Addressed review comments
Addressed all review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fix code format
resolved code format issues.

walter-erquinigo commented on lldb/packages/Python/lldbsuite/test/lldbtest.py:
```diff
@@ -1572,6 +1572,15 @@ def findBuiltClang(self):
+        if path is not None:
+            return path
+
+        return ""
```
return None instead of an empty string

santhoshe447 replied:
Agreed with this comment.
I would like to modify this function to get the lldb-server/gdbserver based on OS.
Is this fine ?

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,146 @@
+class TestDAP_attachByPortNum(lldbdap_testcase.DAPTestCaseBase):
+    def runTargetProgramOnPort(self, port=None, program=None):
+        server_tool = None
+        if lldbplatformutil.getPlatform() == "linux":
+            server_tool = lldbgdbserverutils.get_lldb_server_exe()
+            if server_tool is None:
+                self.dap_server.request_disconnect(terminateDebuggee=True)
+                self.assertIsNotNone(server_tool, "lldb-server not found.")
+            server_tool += " g localhost:" + port + " "
+        elif lldbplatformutil.getPlatform() == "macosx":
+            server_tool = lldbgdbserverutils.get_debugserver_exe()
+            if server_tool is None:
+                self.dap_server.request_disconnect(terminateDebuggee=True)
+                self.assertIsNotNone(server_tool, "debugserver not found.")
+            server_tool += " --listen localhost:" + port + " "
```
could you move the logic that finds automatically the path to lldb-server or debugserver to lldbdap_testcase? This logic seems to be useful to reuse

santhoshe447 replied:
Agreed with your comment.
I think it would be better to move this into lldbtest.py instead of lldb_testcase,py as it has all utility funtions.

santhoshe447 replied:
1. Moved this piece of code into lldbdap_testcase.py: getBuiltinServerToolWithPortArg(self, port):

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,146 @@
+            self.continue_to_exit()
+
+    @skipIfWindows
+    @skipIfNetBSD  # Hangs on NetBSD as well
```
This "as well" looks weird. Did you test it there in fact?

santhoshe447 replied:
This test case is designed to be verified only on Linux and macOS.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,146 @@
+        self.build_and_create_debug_adaptor()
+        program = self.getBuildArtifact("a.out")
+
+        port = "2345"
```
use a different port for each test, just in case the test runner executes them simultaneously

labath replied:
That's definitely not enough, and it's guaranteed to make the test flaky and/or environment sensitive. Both lldb-server and gdb support race-free port allocations by choosing the ports themselves letting the caller know of the actual chosen port.

Writing this logic in a way that works across server implementations and operating systems is not easy, but we already have the code (several in fact, but I'd recommend reusing the implementation in  `test/API/tools/lldb-server/commandline/TestGdbRemoteConnection.py`).

As an alternative, since what you really just want to test is that the connection is made, you could consider using the mock gdb server implementation in `test/API/functionalities/gdb_remote_client/`. That way, it will be you (the python code) who's opening the ports, and there will again be no races.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -676,6 +676,8 @@ void request_attach(const llvm::json::Object &request) {
   const lldb::pid_t pid =
       GetUnsigned(arguments, "pid", LLDB_INVALID_PROCESS_ID);
+  const auto port = GetUnsigned(arguments, "port", LLDB_INVALID_PORT_NUMBER);
+  llvm::StringRef hostname = GetString(arguments, "hostname");
```
you can here set the default to "localhost", which will simplify the logic below

labath commented on lldb/include/lldb/lldb-defines.h:
```diff
@@ -96,6 +96,7 @@
 #define LLDB_INVALID_QUEUE_ID 0
 #define LLDB_INVALID_CPU_ID UINT32_MAX
 #define LLDB_INVALID_WATCHPOINT_RESOURCE_ID UINT32_MAX
+#define LLDB_INVALID_PORT_NUMBER 0
```
Port numbers aren't an lldb invention, so I don't think we should be putting this into the lldb's public (and forever immutable) API.

labath reviewed (changes requested):
FWIW, I think this feature would be more useful and easier to implement if there was just a single setting called "connection url" or something, which accepted all of the usual lldb connection urls (unix-abstract-connect://, listen://, ...).

Since a simple tcp port connection is definitely going to be the most commonly used connection, you can, if you want, treat a simple number (or `:number`) as an alias to `connect://localhost:number`. lldb-server does something similar.

PS: I'm clicking request changes because I don't believe the test for the feature can go in in this form. I don't consider myself an owner for the rest, so you can consider my comments as suggestions.

walter-erquinigo commented:
Thanks, @labath , for chiming in. I actually agree with all your points.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved all review comments.
All review comments have been addressed, and the attach-by-port tests have been verified on
Linux machine. Although the functionality is intended to support both Linux and macOS, we were
unable to verify it on macOS due to lack of access.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command -resolved format check issues.

walter-erquinigo commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,142 @@
+
+    @skipIfWindows
+    @skipIfNetBSD
+    @skipIfRemote
```
you don't need `@skipIfRemote` anymore

santhoshe447 replied:
Agreed, removed @skipIfRemote.

walter-erquinigo commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -749,9 +752,30 @@ void request_attach(const llvm::json::Object &request) {
+        std::string connect_url("connect://localhost:");
+
+        // If the user has provided hostname other than localhost.
+        if (!hostname.empty() && !hostname.starts_with("localhost")) {
```
wouldn't it just work with `localhost` being passed to `llvm::formatv`?

santhoshe447 replied:
Agreed, fixed this.

clayborg commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -348,6 +348,17 @@
+                  "string"
+                ],
+                "description": "TCP/IP port to attach to. Specifying both pid and port is an error."
+              },
```
Do we want to label this `"gdb-remote-port"` instead of `"port"`? I don't think we have any other process plug-ins that support connecting via a port, so it migth be nice to let people know what this is for?

santhoshe447 replied:
Thank you for the suggestion. I think labeling it as "gdb-remote-port" & "gdb-remote-hostname" makes sense for clarity. Since there are no other process plugins that support connecting via port, this lable will also help users understand its purpose more easily.
We used the attribute names "port" and "hostname" because Microsoft visual studio Code IDE examples use the same terminology (https://code.visualstudio.com/Docs/editor/debugging#_launch-configurations). Please reconsider this suggestion to ensure consistency with widely accepted convention.


![image](https://github.com/llvm/llvm-project/assets/150406203/409b9b42-64fc-46c4-a152-94b3357f0488)

clayborg replied:
I think we should make it clear by prefixing with `"gdb-remote-port"` and `"gdb-remote-hostname"` to make it clear this is setting up the GDB remote debug configuration. It then also means we can hard code the `"gdb-remote"` plug-in name as these settings imply that process plugin. The word `"port"` is too general and we might want to use for connecting to a `lldb-dap` process that is running and listening on another host/port as well.

santhoshe447 replied:
Renamed "port" and "hostname" to "gdb-remote-port" and "gdb-remote-hostname"

clayborg commented on lldb/tools/lldb-dap/package.json:
```diff
@@ -348,6 +348,17 @@
+                ],
+                "description": "TCP/IP port to attach to. Specifying both pid and port is an error."
+              },
+              "hostname": {
```
`"gdb-remote-hostname"` similar to above comment?

clayborg commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -749,9 +752,30 @@ void request_attach(const llvm::json::Object &request) {
+          connect_url = llvm::formatv("connect://{0}:", hostname.data());
+        }
+        connect_url += std::to_string(port);
+        g_dap.target.ConnectRemote(listener, connect_url.c_str(), "gdb-remote",
```
We are hardcoding `"gdb-remote"` here, so it might make sense to either rename the above components or have a `"process-plugin"` attribute that defaults to `"gdb-remote"`?

santhoshe447 replied:
Since we don't support any process plugins other than "gdb-remote", would it be a problem to hardcode "gdb-remote"? Adding a "process-plugin" attribute would be an additional burden to maintain and might confuse users.
If we use default "process-pligin"="gdb-remote" would be ok but let's take opinion from forks.

clayborg replied:
It we prefix the port and hostname settings with `"gdb-remote-port"` and `"gdb-remote-hostname"` then we can hard code this as the settings imply that we use the `"gdb-remote"` plug-in.

clayborg reviewed (commented):
Just a few questions if we want to prefix these settings with `"gdb-remote"` as suggested. Please chime in if you have any opinions on the matter.

labath reviewed (changes requested):
You did not respond to my comment about generalizing this to accept a single "url" parameter. I'm not saying it has to be implemented that way, but I would like to hear your thoughts on it, as it definitely has some advantages (and it sounds like Walter agrees).

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command. #91570
Resolved few review commennts.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
Resolved code format issues.

santhoshe447 commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,142 @@
+
+
+class TestDAP_attachByPortNum(lldbdap_testcase.DAPTestCaseBase):
+    def get_free_port(self):
+        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        s.bind(("", 0))
+        port = s.getsockname()[1]
+        s.close()
+        return port
```
I referred to your previous review comments and made these changes. Instead of using a hardcoded same port number, I used an available free port number to attach.
If my understanding is incorrect, could you please clarify whether you mean to use a random port number? Additional information would be helpful to make the necessary changes.

labath replied:
Fair enough. Let me try to rephrase.

The thing I want to avoid is races in port allocation. Your current algorithm works like this:
1. open a socket
2. use it to get a free port (number) from the OS
3. close the socket
4. pass the port number to lldb-server
5. have lldb-server listen on that port
6. connect to that port

This algorithm is racy for two reasons. One, because somewhere between the steps 3 and 5, someone else could start listening on that port. The second is because there's no sequencing between steps 5 and 6. Even if noone claims that port, if lldb-server just happens to be slower for whatever reason (that tends to happen more than you'd think when you're running many tests in parallel), the test can attempt to connect to the port before it gets opened.

Now, compare this to the algorithm in test/API/tools/lldb-server/commandline/TestGdbRemoteConnection.py (test_named_pipe)
1. lldb-server opens a socket
2. lldb-server uses it to get a free port from the OS
3. lldb-server starts listening on the port
4. lldb-server sends the port number to the test (through the named pipe)
5. the test connects to that port

There's no race here. The socket that's used for getting the port number is the same that's used for listening for incoming connections, so there's no possibility for someone to snatch it from under us. And because the test has to wait for the port number before it can connect (and because the lldb-server does steps 3 and 4 in this order), we can be sure the lldb-server will be listening by the time we connect.

That's not the only way to implement these things, but my suggestion is to do this, because the code is already there, and it shouldn't be too hard to make it usable in your test as well.

santhoshe447 replied:
Thank you very much for taking the time to review my work. I truly appreciate your suggestions. 
Your feedback has been very helpful in improving my work.

I have pushed latest changes as per your suggestions.

santhoshe447 committed:
Merge branch 'llvm:main' into main

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - addressed review comments.
Used named pipe to read debug server listening port.
Since we are only using "gdb-remote" process plugin, changed to "gdb-remote-port" and
"gdb-remote-hostname" from "port" and "hostname"

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved code format issues.

santhoshe447 commented:
> FWIW, I think this feature would be more useful and easier to implement if there was just a single setting called "connection url" or something, which accepted all of the usual lldb connection urls (unix-abstract-connect://, listen://, ...).
> 
> Since a simple tcp port connection is definitely going to be the most commonly used connection, you can, if you want, treat a simple number (or `:number`) as an alias to `connect://localhost:number`. lldb-server does something similar.
> 
> PS: I'm clicking request changes because I don't believe the test for the feature can go in in this form. I don't consider myself an owner for the rest, so you can consider my comments as suggestions.

I apologize for missing your earlier comment. To clarify.
Using the URL "connect://localhost:12345" or "connect://:12345" with the "connect" scheme helps invoke the appropriate function to establish a connection.
Pls ref: https://github.com/llvm/llvm-project/blob/main/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp#L148
$gdb-remote number - would work from the command-line, because lldb is taking care of adding "connect://".

Please let me know if my understating is incorrect.

labath commented:
> > FWIW, I think this feature would be more useful and easier to implement if there was just a single setting called "connection url" or something, which accepted all of the usual lldb connection urls (unix-abstract-connect://, listen://, ...).
> > Since a simple tcp port connection is definitely going to be the most commonly used connection, you can, if you want, treat a simple number (or `:number`) as an alias to `connect://localhost:number`. lldb-server does something similar.
> > PS: I'm clicking request changes because I don't believe the test for the feature can go in in this form. I don't consider myself an owner for the rest, so you can consider my comments as suggestions.
> 
> I apologize for missing your earlier comment. To clarify. Using the URL "connect://localhost:12345" or "connect://:12345" with the "connect" scheme helps invoke the appropriate function to establish a connection. Pls ref: https://github.com/llvm/llvm-project/blob/main/lldb/source/Host/posix/ConnectionFileDescriptorPosix.cpp#L148 $gdb-remote number - would work from the command-line, because lldb is taking care of adding "connect://".
> 
> Please let me know if my understating is incorrect.

Your understanding of the code is correct. What I am suggesting is basically to emulate this command line behavior in vscode. 

Basically, instead of two host+port settings, you could have one setting with a sufficiently vague name ("connection url", "remote address", ...) to cover the expanded scope. Then, you implementation could take the user-provided string and fill in the blanks:
- if the user types "1234", you'd convert it to `connect://localhost:1234`
- if the user types "my-host:1234", you'd convert it to `connect://localhost:1234`
- if the user types "unix-abstract-connect://..." (or any other scheme supported by lldb), you'd pass the value unmodified

The advantage of this is that you'd automatically support every connection protocol that lldb understands with almost no effort. If we stick to the current implementation, then if someone later needs to connect through unix sockets (or whatever), he would have to go back and add *another* setting to support connecting through that.

Now, I do see the appeal of having an explicit host/port fields (they're much easier to understand for the average user), which is why I'm not insisting on this approach. Nonetheless, I think this is an option worth considering, and I think we could avoid most of the confusion by carefully crafting the description of this field to guide the user to the most common scenarios.

labath commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,202 @@
+
+        port = 0
+        args = [program]
+        debug_server_tool = self.getBuiltinDebugServerTool()
+        self.process = self.spawnSubprocess(
+            debug_server_tool, args, install_remote=False
+        )
```
same question here, and in the test below.

labath replied:
And everything I said above applies doubly here, since there isn't even a theoretical chance that the two can connect, as they just don't know about each other.

santhoshe447 commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,202 @@
+import select
+
+
+# A class representing a pipe for communicating with debug server.
+# This class includes menthods to open the pipe and read the port number from it.
+class Pipe(object):
+    def __init__(self, prefix):
+        self.name = os.path.join(prefix, "stub_port_number")
+        os.mkfifo(self.name)
+        self._fd = os.open(self.name, os.O_RDONLY | os.O_NONBLOCK)
+
+    def finish_connection(self, timeout):
+        pass
+
+    def read(self, size, timeout):
+        (readers, _, _) = select.select([self._fd], [], [], timeout)
+        if self._fd not in readers:
+            raise TimeoutError
+        return os.read(self._fd, size)
+
+    def close(self):
+        os.close(self._fd)
```
I agree but the specific class is defined based on the host platform. If the host platform is windows, it uses the same Pipe class with different APIs, and for other platforms, it behaves differently.
Therefore, I did not place Pipe class in a common location.


Ref: lldb/test/API/tools/lldb-server/commandline/TestGdbRemoteConnection.py

santhoshe447 replied:
If I make it common it will create more confusion.

labath replied:
I'm not sure I follow you're reasoning (please elaborate), but I'm pretty sure I don't agree with your conclusion.

It is true that the Pipe class is defined differently on different (host) platforms. It's also true that it's interface is more complicated than if it was designed to be used exclusively on (e.g.) posix system. However, I don't think any of this is a reason to not make it common code. However, I don't think any of this is a reason to *not* make it common code. If anything, I would say it's the opposite. Abstracting platform-specific behavior behind common interfaces is the best thing we can do in these circumstances. Sometimes that means the common interface will be a bit more complicated but that's just the reality we live in.

In this case, somebody (disclaimer: it was me, several years ago) already created a common interface and implemented it on both platforms. I don't see a reason to not make use of that. There is no fundamental reason why this test of yours cannot work on windows. If someday, someone cared enough to try to make it work, he would have to (re)implement the pipe abstraction to make it run. Why make him do that if we can make use of an abstraction that's already available?

I'm sorry, but I just don't understand what kind of confusion you're referring to. I don't know if it helps, but I can certainly imagine changing the class name to something less generic (e.g. `HostPipe` ?)

santhoshe447 replied:
I apologize for the confusion regarding the "Pipe " class.
There is a conditional declaration of the "Pipe" class, one for windows and another for non-windows host platform. My confusion was whether I should move both of them or not.
Now I have moved both the "Pipe" classes into common location "lldbgdbserverutils.py" and pushed it.

labath replied:
Thanks. Glad we got that resolved.

santhoshe447 commented on lldb/test/API/tools/lldb-dap/attach/TestDAP_attachByPortNum.py:
```diff
@@ -0,0 +1,202 @@
+        """
+        Tests attaching to a process by process ID and port number.
+        """
+        self.build_and_create_debug_adaptor()
+        program = self.getBuildArtifact("a.out")
+
+        debug_server_tool = self.getBuiltinDebugServerTool()
+        pipe = self.get_debug_server_pipe()
+        args = self.get_debug_server_command_line_args()
+        args += [program]
+        args += ["--named-pipe", pipe.name]
+
+        self.process = self.spawnSubprocess(
+            debug_server_tool, args, install_remote=False
+        )
+
+        # Read the port number from the debug server pipe.
+        port = pipe.read(10, self.default_timeout)
+        # Trim null byte, convert to int
+        port = int(port[:-1])
```
Yes, you are right but I winna test with complete cycle.
I will change this if its required.

santhoshe447 replied:
Even if I remove that it does not impact much.

labath replied:
I would really like to remove it as I don't think it contributes in any way to the test (and it's always better to have less moving parts). Like, even if lldb-dap somehow still ends up connecting to the lldb-server port (in addition to sending the error message), we would never notice it, as the test does not assert the lldb-server state in any way.

If you really wanted to be thorough, a better check would be to open a port (in python) and then verify we do *not* attempt to connect to that port.

santhoshe447 replied:
Yeah, agreed with your point.

clayborg reviewed (commented):
The code looks good, a few nits in comments. 

@labath: did we resolve the port race condition testing issue? We really  don't want this to be a flaky test on overloaded systems. It would be nice to make sure this is solid before getting this in.

santhoshe447 committed:
Merge branch 'llvm:main' into main

labath commented:
> The code looks good, a few nits in comments.
> 
> @labath: did we resolve the port race condition testing issue? We really don't want this to be a flaky test on overloaded systems. It would be nice to make sure this is solid before getting this in.

The current version of the code should be fine. We just need to sort out the code placement issue.

santhoshe447 committed:
Merge branch 'llvm:main' into main

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - fixed review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command - resolved review comments.

Commit:
[lldb-dap] Added "port" property to vscode "attach" command.
I have shifted "Pipe" class to common location "lldb/packages/Python/lldbsuite/test/tools/lldb-server/lldbgdbserverutils.py"

santhoshe447 commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -672,9 +672,14 @@ void request_attach(const llvm::json::Object &request) {
+  const auto gdb_remote_port =
+      GetUnsigned(arguments, "gdb-remote-port", invalid_port);
+  llvm::StringRef gdb_remote_hostname =
+      GetString(arguments, "gdb-remote-hostname", "localhost");
```
Using 127.0.0.1 can provide slight advantage in terms of consistency and compatibility in specific scenarios.
Any suggestions from others?

labath replied:
There are environments (Google's (default) test environment is like that) that don't have an IPv4 stack configured (just v6), so `localhost` would be an advantage there. That said, this definitely wouldn't be the first instance of `127.0.0.1 in lldb.

FWIW, I'm of the opinion that anyone who configures `localhost` to point to something other than their local host deserves whatever is coming their way,

labath reviewed (approved):
My concerns have been addressed (*), so I'm removing my request for changes. It looks like the only way to do that is to Approve the PR, but I'd suggest getting another approval from a lldb-dap owner as well.

(*) There is still the question of the single "url" setting (to rule them all), but I don't consider myself an authority there.

walter-erquinigo commented:
I'll review this today or tomorrow. Thanks for all the activity!

walter-erquinigo reviewed (approved):
This LGTM

omjavaid commented:
This breaks   lldb-api :: tools/lldb-server/commandline/TestGdbRemoteConnection.py on lldb-aarch64-windows bot.

https://lab.llvm.org/buildbot/#/builders/141/builds/376

tedwoodward commented:
> This breaks lldb-api :: tools/lldb-server/commandline/TestGdbRemoteConnection.py on lldb-aarch64-windows bot.
> 
> https://lab.llvm.org/buildbot/#/builders/141/builds/376

  File "C:\Users\tcwg\llvm-worker\lldb-aarch64-windows\llvm-project\lldb\packages\Python\lldbsuite\test\tools\lldb-server\lldbgdbserverutils.py", line 1045, in __init__
    self.name = "lldb-" + str(random.randrange(1e10))
                              ^^^^^^
NameError: name 'random' is not defined


Looks like "import random" should have been added to lldbgdbserverutils.py when the pipe class was moved from TestGDBRemoteConnection.py.
